File: 백준/Gold/1753. 최단경로/최단경로.cc
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

typedef pair<int, int> p;

int N, M, S;

//다익스트라
vector<p> graph[20001];
int cost[20001] = { 0 };
int INF = 987654321;
//가중치, 인덱스
priority_queue<p, vector<p>, greater<p>> pq;

void Insert_And_Init() {
	int s, e, c;

	cin >> N >> M >> S;
	for (int i = 0; i < M; i++){
		cin >> s >> e >> c;
		graph[s].push_back(make_pair(e, c));
	}
	for (int i = 1; i <= N; i++) cost[i] = INF;
	cost[S] = 0;
}

void dijkstra() {
	pq.push(make_pair(0, S));

	while (!pq.empty()) {
		p cur = pq.top(); pq.pop();
		int dist = cur.first; int idx = cur.second;

		//선택된게 어차피 최단경로가 아닌 경우
		//이거 때문에 visited가 필요가 없음
		if (dist > cost[idx]) continue;

		for (int i = 0; i < graph[idx].size(); i++) {
			//반복문안에서 보고 있는 노드
			int nextIdx = graph[idx][i].first;
			//현재 큐에서 선택된 노드까지의 거리 + 직진거리
			int nextDist = dist + graph[idx][i].second;

			//지금 본 경로가 더 가까우면 갱신
			if (cost[nextIdx] > nextDist) {
				cost[nextIdx] = nextDist;
				pq.push(make_pair(nextDist, nextIdx));
			}
		}
	}
}

void Caluclate() {
	dijkstra();
	for (int i = 1; i <= N; i++){
		if (cost[i] == INF) cout << "INF\n";
		else cout << cost[i] << "\n";
	}
}

int main() {
	ios_base::sync_with_stdio(false); cin.tie(NULL);

	Insert_And_Init();

	Caluclate();
}

File: 백준/Gold/1753. 최단경로/최단경로.kt
import java.util.PriorityQueue
import kotlin.math.min

var V = 0
var E = 0
var K = 0

// 다익스트라
const val INF = 987654321
val graph = Array(20001) { mutableListOf<Node>() }
val cost = IntArray(20001) { INF }

data class Node(
    val idx: Int,
    val weight : Int
)

fun insertAndInit() {

    val input = readln().split(" ").map { it.toInt() }
    V = input[0]; E = input[1]
    K = readln().toInt()

    // 간선 입력
    repeat(E) {
        val (u, v, w) = readln().split(" ").map { it.toInt() }
        graph[u].add(Node(v, w))
    }

    cost[K] = 0
}

fun dijkstra() {

    // 우선순위 큐
    val pq = PriorityQueue<Node> { a, b ->
         a.weight - b.weight
    }

    // 자기 자신은 거리 0
    pq.add(Node(K, 0))

    // 큐가 빌때까지
    while (!pq.isEmpty()) {

        // 거리가 제일 가까운 노드 하나 꺼내기
        val curNode = pq.remove()

        // 알아보기 쉽게 거리와 노드 번호 분리
        val dist = curNode.weight
        val idx = curNode.idx

        // 현재 뽑은 거리가 최단경로가 아니면 다음 간선 보기
        if (dist > cost[idx]) continue

        // 현재 노드에 연결된 선분을 조회하며
        graph[idx].forEachIndexed { index, value ->

            // 상대방 노드
            val nextIdx = value.idx

            // idx를 경유하는 상대방까지의 거리
            val nextDist = dist + value.weight

            // 최단경로면 갱신
            if (cost[nextIdx] > nextDist) {
                cost[nextIdx] = nextDist
                pq.add(Node(nextIdx, nextDist))
            }
        }
    }

}

fun calculate() {

    // 다익스트라 실행
    dijkstra()

    // 최단 경로 출력
    repeat(V) {
        if (cost[it + 1] == INF) {
            println("INF")
        } else {
            println(cost[it + 1])
        }
    }
}


fun main() = with(System.`in`.bufferedReader()) {

    insertAndInit()

    calculate()
}